game_catalog/views.py

```diff
@@ -1,3 +1,137 @@
+
+
+class GenreViewSet(viewsets.ModelViewSet):
+    queryset = Genre.objects.all()
+    serializer_class = GenreSerializer
+    permission_classes = [IsAdminOrReadOnly]
+
+
+class StudioViewSet(viewsets.ModelViewSet):
+    queryset = Studio.objects.all()
+    serializer_class = StudioSerializer
+    permission_classes = [IsAdminOrReadOnly]
+
+
+class GameViewSet(viewsets.ModelViewSet):
+    queryset = Game.objects.all()
+    permission_classes = [IsAdminOrReadOnly]
+
+    def get_serializer_class(self):
+        if self.action in ["create", "update", "partial_update"]:
+            print("self.action in create, update")
+            return GameWriteSerializer
+        return GameSerializer
+
+    @action(
+        detail=True,
+        methods=["get", "post"],
+        permission_classes=[IsAuthenticatedOrReadOnly],
+    )
+    def comments(self, request, pk=None):
+        game = self.get_object()
+        if request.method == "POST":
+            serializer = GameCommentSerializer(data=request.data)
+            if serializer.is_valid():
+                serializer.save(user=request.user, game=game)
+                return Response(serializer.data, status=status.HTTP_201_CREATED)
+            return Response(serializer.errors, status=status.HTTP_400_BAD_REQUEST)
+
+        comments = game.comments.all()
+        serializer = GameCommentSerializer(comments, many=True)
+        return Response(serializer.data)
+
+    @action(detail=True, methods=["post"], permission_classes=[IsAuthenticated])
+    def add_to_favorites(self, request, pk=None):
+        game = self.get_object()
+        user = request.user
+        if game in user.favorite_games.all():
+            return Response(
+                {"detail": "Game is already in favorites."},
+                status=status.HTTP_400_BAD_REQUEST,
+            )
+
+        user.favorite_games.add(game)
+        return Response(
+            {"detail": "Game added to favorites."}, status=status.HTTP_200_OK
+        )
+
+    @action(detail=True, methods=["post"], permission_classes=[IsAuthenticated])
+    def remove_from_favorites(self, request, pk=None):
+        game = self.get_object()
+        user = request.user
+        if game not in user.favorite_games.all():
+            return Response(
+                {"detail": "Game is not in favorites."},
+                status=status.HTTP_400_BAD_REQUEST,
+            )
+
+        user.favorite_games.remove(game)
+        return Response(
+            {"detail": "Game removed from favorites."}, status=status.HTTP_200_OK
+        )
+
+
+class CommentViewSet(viewsets.ModelViewSet):
+    queryset = Comment.objects.all()
+    serializer_class = CommentSerializer
+    permission_classes = [IsAuthenticated, IsOwnerOrAdmin]
+
+    def perform_create(self, serializer):
+        serializer.save(user=self.request.user)
+
+
+class UserViewSet(viewsets.ModelViewSet):
+    queryset = CustomUser.objects.all()
```

Comment:
did you test what are you getting here and don't you again have n+1 problem here?

Reply:
Yes, we have N+1 problem here.

So i added `prefetch_related `for queryset in `UserViewSet `and remove explicit field definition in User Serializers for favorite_games ( `favorite_games = GameSerializer(many=True, read_only=True)`)
And now instead of serialized Game objects (as with GameSerializer), you'll see a list of game IDs.

b8397c21707cc19c3a90c71b476d349b0801fd5c

Reply:
minor fix

d93e3828c90c3736153f47d03d59a848e922cbc5